File: app/src/main/java/com/ubiqsmart/utils/AddressNameConverter.java
package com.ubiqsmart.utils;

import android.content.Context;
import com.ubiqsmart.repository.data.Wallet;

import java.io.*;
import java.util.*;

public class AddressNameConverter {

  private HashMap<String, String> mapdb;
  private static AddressNameConverter instance;

  public static AddressNameConverter getInstance(Context context) {
    if (instance == null) {
      instance = new AddressNameConverter(context);
    }
    return instance;
  }

  private AddressNameConverter(Context context) {
    try {
      load(context);
    } catch (Exception e) {
      mapdb = new HashMap<>();
    }
  }

  public synchronized void put(String address, String name, Context context) {
    if (name == null || name.length() == 0) {
      mapdb.remove(address);
    } else {
      mapdb.put(address, name.length() > 22 ? name.substring(0, 22) : name);
    }
    save(context);
  }

  public String get(String addresse) {
    return mapdb.get(addresse);
  }

  public boolean contains(final String address) {
    return mapdb.containsKey(address);
  }

  public List<Wallet> getAsAddressBook() {
    final List<Wallet> erg = new ArrayList<>();
    for (Map.Entry<String, String> entry : mapdb.entrySet()) {
      erg.add(new Wallet(entry.getValue(), entry.getKey()));
    }
    Collections.sort(erg);
    return erg;
  }

  public synchronized void save(Context context) {
    try {
      FileOutputStream fout = new FileOutputStream(new File(context.getFilesDir(), "namedb.dat"));
      ObjectOutputStream oos = new ObjectOutputStream(fout);
      oos.writeObject(mapdb);
      oos.close();
      fout.close();
    } catch (Exception ignored) {
    }
  }

  @SuppressWarnings("unchecked") public synchronized void load(Context context) throws IOException, ClassNotFoundException {
    final FileInputStream fout = new FileInputStream(new File(context.getFilesDir(), "namedb.dat"));
    final ObjectInputStream oos = new ObjectInputStream(new BufferedInputStream(fout));
    mapdb = (HashMap<String, String>) oos.readObject();
    oos.close();
    fout.close();
  }

}

File: app/src/main/java/com/ubiqsmart/app/utils/AddressNameConverter.kt
package com.ubiqsmart.app.utils

import android.content.Context
import com.ubiqsmart.domain.models.WalletEntry

import java.io.*
import java.util.*

@Deprecated("Replace with room entities and repository")
class AddressNameConverter private constructor(context: Context) {

  private var mapdb: HashMap<String, String>? = null

  private val context: Context = context.applicationContext

  val asAddressBook: List<WalletEntry>
    get() {
      val erg = ArrayList<WalletEntry>()
      for ((key, value) in mapdb!!) {
        erg.add(WalletEntry(value, key))
      }
      Collections.sort(erg)
      return erg
    }

  init {
    try {
      load()
    } catch (e: Exception) {
      mapdb = HashMap()
    }

  }

  @Synchronized
  fun put(address: String, name: String?) {
    if (name == null || name.isEmpty()) {
      mapdb!!.remove(address)
    } else {
      mapdb!!.put(address, if (name.length > 22) name.substring(0, 22) else name)
    }
    save()
  }

  operator fun get(address: String): String? {
    return mapdb!![address]
  }

  operator fun contains(address: String): Boolean {
    return mapdb!!.containsKey(address)
  }

  @Synchronized
  private fun save() {
    try {
      val fout = FileOutputStream(File(context.filesDir, "namedb.dat"))
      val oos = ObjectOutputStream(fout)
      oos.writeObject(mapdb)
      oos.close()
      fout.close()
    } catch (ignored: Exception) {
    }

  }

  @Synchronized
  @Throws(IOException::class, ClassNotFoundException::class)
  private fun load() {
    val fout = FileInputStream(File(context.filesDir, "namedb.dat"))
    val oos = ObjectInputStream(BufferedInputStream(fout))
    mapdb = oos.readObject() as HashMap<String, String>
    oos.close()
    fout.close()
  }

  companion object {

    private var instance: AddressNameConverter? = null

    fun getInstance(context: Context): AddressNameConverter {
      return instance ?: AddressNameConverter(context).also { instance = it }
    }
  }

}
